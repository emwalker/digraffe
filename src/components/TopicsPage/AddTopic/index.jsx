// @flow
import React, { Component } from 'react'
import { Button, FormGroup, Input } from 'reactstrap'

import createTopicMutation from '../../../mutations/createTopicMutation'

type Props = {
  organization: {
    id: string,
    resourceId: string,
  },
  relay: {
    environment: Object,
  },
}

type State = {
  name: string,
}

class AddTopic extends Component<Props, State> {
  state = {
    name: '',
  }

  onSubmit = () => {
    const {
      id: orgId,
      resourceId: organizationId,
    } = this.props.organization
    createTopicMutation(
      this.props.relay.environment,
      orgId,
      {
        organizationId,
        name: this.state.name,
      },
    )
    this.setState({ name: '' })
  }

  updateName = (event: Object) => {
    this.setState({ name: event.currentTarget.value })
  }

  render = () => (
    <div className="form-container">
      <FormGroup>
        <Input
          className="topic-name test-topic-name"
          placeholder="Topic name"
          value={this.state.name}
          onChange={this.updateName}
        />
        <Button size="sm" onClick={this.onSubmit}>Add</Button>
      </FormGroup>
    </div>
  )
}

export default AddTopic
