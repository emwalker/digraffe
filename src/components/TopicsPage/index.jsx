// @flow
import React from 'react'
import { graphql, createFragmentContainer } from 'react-relay'
import { isEmpty } from 'ramda'

import List from '../ui/List'
import Topic from '../ui/Topic'
import { liftNodes } from '../../utils'

type Props = {
  view: {
    topics: Object,
  },
}

const TopicsPage = ({ view, ...props }: Props) => {
  const topics = liftNodes(view.topics)
  return (
    <div>
      <List
        placeholder="There are no topics"
        hasItems={!isEmpty(topics)}
      >
        { topics.map(topic => (
          <Topic
            key={topic.resourcePath}
            topic={topic}
            {...props}
          />
        )) }
      </List>
    </div>
  )
}

export const query = graphql`
query TopicsPage_query_Query($orgIds: [ID!]) {
  view(organizationIds: $orgIds) {
    ...TopicsPage_view
  }
}`

export default createFragmentContainer(TopicsPage, graphql`
  fragment TopicsPage_view on View {
    topics(first: 1000) @connection(key: "View_topics") {
      edges {
        node {
          resourcePath
          ...Topic_topic
        }
      }
    }
  }
`)
