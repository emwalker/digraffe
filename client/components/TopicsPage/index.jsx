// @flow
import React from 'react'
import { compose, isNil, map, prop, propOr, reject } from 'ramda'
import { graphql, createFragmentContainer } from 'react-relay'
import { ListGroup, ListGroupItem } from 'reactstrap'

import AddTopic from './AddTopic'

const topicList = compose(reject(isNil), map(prop('node')), propOr([], 'edges'))

type Props = {
  organization: {
    topics: Object,
  },
  relay: {
    environment: Object,
  }
}

const TopicsPage = ({ organization, relay }: Props) => (
  <div>
    <h1>Topics</h1>
    <div className="row">
      <div className="col">
        <ListGroup>
          {topicList(organization.topics).map(({ id, name, resourcePath }) => (
            <ListGroupItem key={id} tag="a" href={resourcePath}>{name}</ListGroupItem>
          ))}
        </ListGroup>
      </div>
      <div className="col-5">
        <AddTopic
          className="test-add-topic"
          organization={organization}
          relay={relay}
        />
      </div>
    </div>
  </div>
)

export default createFragmentContainer(TopicsPage, graphql`
  fragment TopicsPage_viewer on User {
    name
  }

  fragment TopicsPage_organization on Organization {
    id
    resourceId

    topics(first: 100) @connection(key: "Organization_topics") {
      edges {
        node {
          id
          name
          resourcePath
          description
        }
      }
    }
  }
`)
