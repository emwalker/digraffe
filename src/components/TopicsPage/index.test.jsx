import React from 'react'
import { shallow } from 'enzyme'
import TopicsPage from './index'

jest.mock('react-relay', () => ({ createFragmentContainer: component => component }))

describe('<TopicsPage />', () => {
  const viewer = {
    name: 'Rezrov',
  }

  const view = {
    resourceId: 'organization:tyrell',
    topics: {
      edges: [
        {
          node: {
            name: 'Frotz',
            resourcePath: '/topics/1234',
          },
        },
      ],
    },
  }

  const wrapper = shallow(
    <TopicsPage
      view={view}
      viewer={viewer}
    />,
  )

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
