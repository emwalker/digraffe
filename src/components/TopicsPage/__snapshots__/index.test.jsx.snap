// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopicsPage /> renders 1`] = `
<div>
  <Component
    hasItems={true}
    placeholder="There are no topics"
  >
    <Topic
      key="/topics/1234"
      topic={
        Object {
          "name": "Frotz",
          "resourcePath": "/topics/1234",
        }
      }
      viewer={
        Object {
          "name": "Rezrov",
        }
      }
    />
  </Component>
</div>
`;
